core/toolbox/category.js

```diff
@@ -0,0 +1,826 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview A toolbox category used to organize blocks in the toolbox.
+ * @author aschmiedt@google.com (Abby Schmiedt)
+ */
+'use strict';
+
+goog.provide('Blockly.ToolboxCategory');
+
+goog.require('Blockly.CollapsibleToolboxItem');
+goog.require('Blockly.registry');
+goog.require('Blockly.utils.aria');
+goog.require('Blockly.utils.object');
+goog.require('Blockly.utils.toolbox');
+
+goog.requireType('Blockly.ToolboxItem');
+
+
+/**
+ * Class for a category in a toolbox.
+ * @param {!Blockly.utils.toolbox.CategoryJson} categoryDef The information needed
+ *     to create a category in the toolbox.
+ * @param {!Blockly.IToolbox} toolbox The parent toolbox for the category.
+ * @param {Blockly.ToolboxCategory=} opt_parent The parent category or null if
+ *     the category does not have a parent.
+ * @constructor
+ * @extends {Blockly.CollapsibleToolboxItem}
+ */
+Blockly.ToolboxCategory = function(categoryDef, toolbox, opt_parent) {
+  Blockly.ToolboxCategory.superClass_.constructor.call(
+      this, categoryDef, toolbox);
+
+  /**
+   * The name that will be displayed on the category.
+   * @type {string}
+   * @protected
+   */
+  this.name_ = Blockly.utils.replaceMessageReferences(categoryDef['name']);
+
+  var categories = categoryDef['contents'].filter(function(item) {
+    return item['kind'].toUpperCase() == 'CATEGORY';
+  });
+
+  /**
+   * True if this category has subcategories, false otherwise.
+   * @type {boolean}
+   * @private
+   */
+  this.hasChildren_ = !!categories.length;
+
+  /**
+   * The parent of the category.
+   * @type {?Blockly.ToolboxCategory}
+   * @protected
+   */
+  this.parent_ = opt_parent || null;
+
+  /**
+   * The level that the category is nested at.
+   * @type {number}
+   * @protected
+   */
+  this.level_ = this.parent_ ? this.parent_.getLevel() + 1 : 0;
+
+  /**
+   * The colour of the category.
+   * @type {string}
+   * @protected
+   */
+  this.colour_ = this.getColour_(categoryDef);
+
+  /**
+   * The html container for the category.
+   * @type {?Element}
+   * @protected
+   */
+  this.htmlDiv_ = null;
+
+  /**
+   * The html element for the category row.
+   * @type {?Element}
+   * @protected
+   */
+  this.rowDiv_ = null;
+
+  /**
+   * The html elmeent that holds children of the category row.
+   * @type {?Element}
+   * @protected
+   */
+  this.rowContents_ = null;
+
+  /**
+   * The html element for the toolbox icon.
+   * @type {?Element}
+   * @protected
+   */
+  this.iconDom_ = null;
+
+  /**
+   * Container for any children categories.
+   * @type {?Element}
+   * @protected
+   */
+  this.subcategoriesDiv_ = null;
+
+  /**
+   * All the css class names that are used to create a category.
+   * @type {!Blockly.ToolboxCategory.CssConfig}
+   * @protected
+   */
+  this.cssConfig_ = {
```

Comment:
Suggestion: make a helper to generate this ('makeDefaultCssConfig_`?)

You could either leave the mixin code here or put it into the helper--I haven't thought through the implications of each choice.